### ISSUE TITLE
autoconf detection of libbfd can not  work

### ISSUE THREAD

**Issue by olafhering** [ocaml/ocaml#9297]
Since the introduction of `configure.ac`, the test for libbfd can not work. `AC_CHECK_LIB` caches the result. If the first attempt fails, further attempts are considered as failed upfront as well.

My workaround to get a `ocamlobjinfo` for `cmxs` files looks like that.

````diff
--- a/configure.ac
+++ b/configure.ac
@@ -1565,7 +1565,7 @@ AS_IF([test x"$enable_graph_lib" = "xno" ],
 
 AC_CHECK_HEADER([bfd.h],
   [libbfd_ling=""
-  AC_CHECK_LIB([bfd], [bfd_openr], [libbfd_link="-lbfd"])
+  AC_CHECK_LIB([bfd], [bfd_openr], [libbfd_link="-lbfd -ldl -liberty -lz"], [], [-ldl -liberty -lz])
   AS_IF([test -z "$libbfd_link"],
     [AC_CHECK_LIB([bfd], [bfd_openr], [libbfd_link="-lbfd -ldl"], [], [-ldl])])
   AS_IF([test -z "$libbfd_link"],
````
I see 4.10 has some changes, the patch above is for 4.08. But in fact 4.10 fails as well for me.

**Comment by Octachron** [comment]
cc @shindere

**Comment by olafhering** [comment]
I looked through `config.log` from 4.10. It seems '-ldl' is missing.
`-lbfd -liberty -lm` -> `undefined reference to 'dlopen'/'inflateInit_'/'compressBound'`
`-lbfd -liberty -lz -lm` -> `undefined reference to 'dlopen'`
`-lbfd -liberty -lz -lintl -lm` -> `cannot find -lintl`

**Comment by shindere** [comment]
The initial problem you report has been resolved, at least on trunk, cf.
the `unset ac_cv_lib_bfd_bfd_openr`lines in `configure.ac`.

Regarding the use of the `-ldl` flag, I don't really know yet what the
right approach would be. Should it be added from the beginning? The
point is that it works without it on our CI machines, I assume because
static versions of the required libraries are available.

**Comment by olafhering** [comment]
At least in Leap 15.1 it fails. Static libraries exists:
````shell
# rpm -ql binutils-devel | grep -vw h$
/usr/lib64/libbfd.a
/usr/lib64/libiberty.a
/usr/lib64/libopcodes.a
````

````
configure:16122: checking for bfd_openr in -lbfd
configure:16147: /usr/bin/gcc -o conftest -fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection     conftest.c -lbfd -liberty -lz -lm  >&5
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /usr/lib64/gcc/x86_64-suse-linux/7/../../../../lib64/libbfd.a(plugin.o): in function `try_load_plugin':
/home/abuild/rpmbuild/BUILD/binutils-2.32/build-dir/bfd/../../bfd/plugin.c:243: undefined reference to `dlopen'
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /home/abuild/rpmbuild/BUILD/binutils-2.32/build-dir/bfd/../../bfd/plugin.c:271: undefined reference to `dlsym'
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /home/abuild/rpmbuild/BUILD/binutils-2.32/build-dir/bfd/../../bfd/plugin.c:256: undefined reference to `dlclose'
/usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /home/abuild/rpmbuild/BUILD/binutils-2.32/build-dir/bfd/../../bfd/plugin.c:246: undefined reference to `dlerror'
collect2: error: ld returned 1 exit status
````

**Comment by shindere** [comment]
I'm unsure who exactly needs e.g. `dlopen` and why it is really
required. Is that to say that libbfd itself has kind of a plugin-based
mechanism that will want to load dynamic parts even when one uses only
static libraries?

**Comment by olafhering** [comment]
I do not know the internals of binutils, but the name `try_load_plugin` suggests that a module system exists. How are modules supposed to be loaded if not with `dlopen`?

**Comment by shindere** [comment]
Olaf Hering (2020/02/11 16:49 +0000):
> I do not know the internals of binutils, but the name
> `try_load_plugin` suggests that a module system exists. How are
> modules supposed to be loaded if not with `dlopen`?

I don't know -- it works without it, here...

**Comment by shindere** [comment]
If you could do something like

```bash
./configure BFD_LDFLAGS=-ld
```

would that be okay?

Note that this does not work at the moment, I am just asking to see
whether that would be good enough.

**Comment by olafhering** [comment]
I briefly browsed binutils sources. Their configure has a `--enable-plugins`. Was the binutils, which is used in CI, built with plugin support?

What is wrong with using `-ldl`, why was it removed? Why would I need to apply workarounds to fix this regression?

**Comment by shindere** [comment]
Actually: I didn't realise, until now, that there actually was a
regression. My bad, I'm sorry about this. Working on a fix, will send
link to the PR as soon as it's ready and would appreciated it a lot if
you can give it a try.

**Comment by shindere** [comment]
@olafhering: I just opened #9302 to fix this issue, with my apologies.

Would you please mind giving the code a try and let us know whether it
fixes your problem?

**Comment by Octachron** [comment]
Closed by #9311 .
If you think that the support needs to be backported, please don't hesitate to mention it.

**Comment by olafhering** [comment]
Well, in case there will be a 4.09.1 it should be part of the changes to be backported to such bugfix release.
